src/stan/mcmc/hmc/xhmc/base_xhmc.hpp

```diff
@@ -0,0 +1,223 @@
+            valid_subtree
+              = build_tree(this->depth_, z_propose,
+                           sum_numer_subtree, sum_weight_subtree,
+                           H0, -1, n_leapfrog, sum_metro_prob, writer);
+            z_minus.ps_point::operator=(this->z_);
+          }
+
+          sum_numer += sum_numer_subtree;
+          sum_weight += sum_weight_subtree;
+          if (!valid_subtree) break;
+
+          // Sample from an accepted subtree
+          ++(this->depth_);
+
+          double accept_prob = sum_weight_subtree / sum_weight;
+          if (this->rand_uniform_() < accept_prob)
+            z_sample = z_propose;
+
+            // Break if exhaustion criterion is satisfied
+            if (std::fabs(sum_numer / sum_weight) < x_delta_)
+              break;
+        }
+
+        this->n_leapfrog_ = n_leapfrog;
+
+        // Compute average acceptance probabilty across entire trajectory,
+        // even over subtrees that may have been rejected
+        double accept_prob
+          = sum_metro_prob / static_cast<double>(n_leapfrog + 1);
+
+        this->z_.ps_point::operator=(z_sample);
+        this->energy_ = this->hamiltonian_.H(this->z_);
+        return sample(this->z_.q, -this->z_.V, accept_prob);
+      }
+
+      void get_sampler_param_names(std::vector<std::string>& names) {
+        names.push_back("stepsize__");
+        names.push_back("treedepth__");
+        names.push_back("n_leapfrog__");
+        names.push_back("divergent__");
+        names.push_back("energy__");
+      }
+
+      void get_sampler_params(std::vector<double>& values) {
+        values.push_back(this->epsilon_);
+        values.push_back(this->depth_);
+        values.push_back(this->n_leapfrog_);
+        values.push_back(this->divergent_);
+        values.push_back(this->energy_);
+      }
+
+      // Returns number of valid points in the completed subtree
+      int build_tree(int depth, ps_point& z_propose,
+                     double& sum_numer, double& sum_weight,
+                     double H0, double sign, int& n_leapfrog,
+                     double& sum_metro_prob,
+                     interface_callbacks::writer::base_writer& writer) {
+        // Base case
+        if (depth == 0) {
+            this->integrator_.evolve(this->z_, this->hamiltonian_,
+                                     sign * this->epsilon_,
+                                     writer);
+            ++n_leapfrog;
+
+            double h = this->hamiltonian_.H(this->z_);
+            if (boost::math::isnan(h))
+              h = std::numeric_limits<double>::infinity();
+
+            if ((h - H0) > this->max_deltaH_) this->divergent_ = true;
+
+            double pi = exp(H0 - h);
+            sum_numer += pi * this->hamiltonian_.dG_dt(this->z_);
+            sum_weight += pi;
+            sum_metro_prob += pi > 1 ? 1 : pi;
+
+            z_propose = this->z_;
+
+            return !this->divergent_;
+
+        } else {
```

Comment:
Don't need the `else` block here given that you returned from the `if` branch.  This is one of the standard refactors that seemed wrong to me at first.   But once you get used to always putting the "special" cases first, it becomes much easier to read the "normal" control flow.